components/remote_settings/src/storage.rs

```diff
@@ -19,37 +21,96 @@ use crate::{Attachment, RemoteSettingsRecord, Result};
-///     to avoid the possibility of SQLITE_BUSY.  Or maybe not, the writes seem like they should be
-///     very fast.
-///   - Maybe we should refactor this to use the DAO pattern like suggest does.
-pub struct Storage {}
+pub struct Storage {
+    conn: Connection,
+}
 
 impl Storage {
-    pub fn new(_path: Utf8PathBuf) -> Result<Self> {
-        Ok(Self {})
+    pub fn new(path: Utf8PathBuf) -> Result<Self> {
+        let conn = Connection::open(path)?;
+        let storage = Self { conn };
+        storage.initialize_database()?;
+
+        Ok(storage)
+    }
+
+    // Create the different tables for records and attachements for every new sqlite path
+    fn initialize_database(&self) -> Result<()> {
+        self.conn.execute_batch(
+            "
+        CREATE TABLE IF NOT EXISTS records (
+            id TEXT PRIMARY KEY,
+            collection_url TEXT NOT NULL,
+            data BLOB NOT NULL
+        );
+        CREATE TABLE IF NOT EXISTS attachments (
+            id TEXT PRIMARY KEY,
+            collection_url TEXT NOT NULL,
+            data BLOB NOT NULL
+        );
+        CREATE TABLE IF NOT EXISTS collection_metadata (
+            collection_url TEXT PRIMARY KEY,
+            last_modified INTEGER,
+            fetched BOOLEAN
+        );
+        ",
+        )?;
+        Ok(())
     }
 
     /// Get the last modified timestamp for the stored records
     ///
     /// Returns None if no records are stored or if `collection_url` does not match the
     /// `collection_url` passed to `set_records`.
-    pub fn get_last_modified_timestamp(&self, _collection_url: &str) -> Result<Option<u64>> {
-        Ok(None)
+    pub fn get_last_modified_timestamp(&self, collection_url: &str) -> Result<Option<u64>> {
+        let mut stmt = self
+            .conn
+            .prepare("SELECT last_modified FROM collection_metadata WHERE collection_url = ?")?;
+        let result: Option<u64> = stmt
+            .query_row((collection_url,), |row| row.get(0))
+            .optional()?;
+        Ok(result)
     }
 
     /// Get cached records for this collection
     ///
-    /// Returns None if no records are stored or if `collection_url` does not match the
-    /// `collection_url` passed to `set_records`.
-    pub fn get_records(&self, _collection_url: &str) -> Result<Option<Vec<RemoteSettingsRecord>>> {
-        Ok(None)
+    /// Returns None if no records are stored or if `collection_url` does not match the `collection_url` passed
+    /// to `set_records`.
+    pub fn get_records(&self, collection_url: &str) -> Result<Option<Vec<RemoteSettingsRecord>>> {
+        // Check if the collection has been fetched before
+        let mut stmt = self
+            .conn
+            .prepare("SELECT fetched FROM collection_metadata WHERE collection_url = ?")?;
+        let fetched: Option<bool> = stmt
+            .query_row(params![collection_url], |row| row.get(0))
+            .optional()?;
```

Comment:
How about using a transaction for this?  I think that means you'll get a consistent read even if another thread is calling `set_records` at the same time.  Maybe none if this matters because of the mutex, but I don't love relying on that.